File: frontend-v2/src/features/chat/lib/mention.ts
// Pure trigger detection / replacement for the composer's @ / mention menu.
// No React, no I/O — unit-testable in isolation.
//
//   resolveTrigger(text, caret) → active trigger at the caret, or null
//   replaceTrigger(text, range, insert) → text with the trigger span replaced
//
// Rules (see task spec):
//   - "@" fires at a word boundary (line start, or after whitespace / an
//     opening bracket / CJK char / punctuation).
//   - "/" fires only at the start of a line or the start of the whole text.
//   - The query runs from the trigger char to the caret; any whitespace inside
//     it invalidates the trigger, and a query longer than 40 chars is dropped.

export type MentionKind = "at" | "slash"

export interface MentionTrigger {
  kind: MentionKind
  /** Query text between the trigger char and the caret (may be empty). */
  query: string
  /** Index of the trigger char ("@" or "/") in `text`. */
  start: number
  /** Caret index (exclusive end of the replaced span). */
  end: number
}

export interface MentionRange {
  start: number
  end: number
}

const MAX_QUERY = 40

// Chars that count as a left boundary for an "@" mention: ASCII punctuation,
// CJK symbols/punctuation (U+3000–303F), fullwidth forms (U+FF00–FFEF), and
// CJK ideographs (U+3400–9FFF). Written with \u escapes to stay ASCII-clean.
const BOUNDARY = /[[({<,.!?;:'"`\u3000-\u303f\u3400-\u9fff\uff00-\uffef]/

function isAtBoundary(text: string, index: number): boolean {
  if (index === 0) return true
  const prev = text[index - 1] ?? ""
  return /\s/.test(prev) || BOUNDARY.test(prev)
}

function isLineStart(text: string, index: number): boolean {
  return index === 0 || text[index - 1] === "\n"
}

export function resolveTrigger(text: string, caret: number): MentionTrigger | null {
  const end = Math.min(Math.max(caret, 0), text.length)

  // Any whitespace between the trigger char and the caret invalidates the
  // query, so the active trigger must sit inside the whitespace-free run that
  // ends at the caret. Find that run's start first.
  let runStart = end
  while (runStart > 0 && !/\s/.test(text[runStart - 1] ?? "")) runStart--

  // Within the run, the *leftmost* trigger char that satisfies its boundary
  // rule wins. This keeps "@src/foo" an "@" mention even though the path holds
  // a "/", instead of letting that inner "/" (which fails the line-start rule)
  // swallow the trigger.
  for (let i = runStart; i < end; i++) {
    const ch = text[i]
    if (ch !== "@" && ch !== "/") continue
    const kind: MentionKind = ch === "@" ? "at" : "slash"
    if (kind === "at" ? !isAtBoundary(text, i) : !isLineStart(text, i)) continue

    const query = text.slice(i + 1, end)
    if (query.length > MAX_QUERY) return null
    return { kind, query, start: i, end }
  }

  return null
}

export function replaceTrigger(
  text: string,
  range: MentionRange,
  insert: string,
): { text: string; caret: number } {
  const before = text.slice(0, range.start)
  const after = text.slice(range.end)
  // Always leave a single trailing space so the next token starts cleanly.
  const trailing = after.startsWith(" ") ? "" : " "
  const chunk = `${insert}${trailing}`
  return { text: `${before}${chunk}${after}`, caret: range.start + chunk.length }
}

File: mobile/lib/features/chat/utils/mention.dart
/// Pure trigger detection / replacement for the composer's `@` / `/` mention
/// menu — a 1:1 port of frontend-v2 `features/chat/lib/mention.ts`:
/// - `@` fires at a word boundary (line start, whitespace, bracket,
///   punctuation, or a CJK char before it);
/// - `/` fires only at the start of a line;
/// - whitespace inside the query invalidates the trigger; queries longer
///   than 40 chars are dropped.
library;

enum MentionKind { at, slash }

class MentionTrigger {
  const MentionTrigger({
    required this.kind,
    required this.query,
    required this.start,
    required this.end,
  });

  final MentionKind kind;
  final String query;
  final int start;
  final int end;

  String get key => '${kind.name}:$start:$query';
}

const _maxQuery = 40;

final _boundary = RegExp(
  '[\\[({<,.!?;:\'"`\\u3000-\\u303f\\u3400-\\u9fff\\uff00-\\uffef]',
);
final _whitespace = RegExp(r'\s');

bool _isAtBoundary(String text, int index) {
  if (index == 0) return true;
  final prev = text[index - 1];
  return _whitespace.hasMatch(prev) || _boundary.hasMatch(prev);
}

bool _isLineStart(String text, int index) =>
    index == 0 || text[index - 1] == '\n';

MentionTrigger? resolveTrigger(String text, int caret) {
  final end = caret.clamp(0, text.length);

  var runStart = end;
  while (runStart > 0 && !_whitespace.hasMatch(text[runStart - 1])) {
    runStart--;
  }

  // Leftmost trigger char in the whitespace-free run wins, so "@src/foo"
  // stays an @-mention despite the inner "/".
  for (var i = runStart; i < end; i++) {
    final ch = text[i];
    if (ch != '@' && ch != '/') continue;
    final kind = ch == '@' ? MentionKind.at : MentionKind.slash;
    final valid =
        kind == MentionKind.at ? _isAtBoundary(text, i) : _isLineStart(text, i);
    if (!valid) continue;
    final query = text.substring(i + 1, end);
    if (query.length > _maxQuery) return null;
    return MentionTrigger(kind: kind, query: query, start: i, end: end);
  }
  return null;
}

({String text, int caret}) replaceTrigger(
  String text,
  MentionTrigger trigger,
  String insert,
) {
  final before = text.substring(0, trigger.start);
  final after = text.substring(trigger.end);
  final trailing = after.startsWith(' ') ? '' : ' ';
  final chunk = '$insert$trailing';
  return (text: '$before$chunk$after', caret: trigger.start + chunk.length);
}
